eth2/utils/hashing/src/lib.rs

```diff
@@ -8,6 +8,31 @@ pub fn hash(input: &[u8]) -> Vec<u8> {
     result
 }
 
+// Get merkle root of some hashed values - the input leaf nodes is expected to already be hashed
+// Outputs a `Vec<u8>` byte array of the merkle root given a set of leaf node values.
+pub fn merkle_root(values: &[Vec<u8>]) -> Vec<u8> {
```

Comment:
There are a few occasions where this algorithm won't function:

- When `values.len() == 0` the function will panic on line 33.
- When `values.len().is_power_of_two() == false` the function will produce an invalid result. (See the comment [here](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#merkle_root))

There's two ways we can deal with this:
1. Add some explicit `assert` statements to make it clear that these are invalid inputs (e.g., `assert!(values.len().is_power_of_two())`
2. Change the return type of the function to `Option<Vec<u8>>` and return `None` if `values.is_empty()` or `!values.len().is_power_of_two()` (and change line 33 to `Some(o[1].clone())`).

I think (2) is the safest in this scenario, so I'd recommend to go ahead with that.

Apologies, I didn't consider this last time we were chatting.